apache/doris-thirdparty
# [Enhancement] Add HDFS TDE/KMS functions. Now only support HDFS3.

## What is the purpose of the change
[Enhancement] Add HDFS TDE/KMS functions. 

## Test Steps with Doris
1. Compile this third-partiry lib with doris.
2. Setup hadoop3 and hive3 env and kms-related env.
3. Start hdfs server, kms server, hive metastore server.
4. Create hive warehouse root dir and create encryption zone as this root dir path on the HDFS.
 ```
 // hive warehouse root dir 
hadoop fs -mkdir /data/hive
 // create key
hadoop key create key1
// create encryption zone as hive warehouse root dir path.
hdfs crypto -createZone -keyName key1 -path /data/hive
```
5. Create hive tables. And you can check the raw encrypted files.
 ```
// can check raw encrypted file
hdfs dfs -cat /.reserved/raw/data/hive/*
```
6. Add kms settings in`be/conf/hdfs-site.xml` or `CREATE RESOURCE` DDL.
`dfs.encryption.key.provider.uri: 'kms://http@localhost:9600/kms'`
7. Use multi-catalog to read hive tables. The hive tables in encryption zone  can be queried.


## Only Support HDFS3 Server now. HDFS2 Server doesn't work
The root cause is HDFS3 Namenode server will return `fileEncryptionInfo` if the file path is under encryption zone when`FileSystem::getFileStatus()` is called. But HDFS2 namenode server doesn't return. It should be able to get similar file encryption information, need more research.
